### PR TITLE
Fix some occurrences of player inventory clear

### PR DESCRIPTION
**Pull request by tornac1234**
- [x] Prevent players to build on other players, or to destroy a building with players inside
> When the RemotePlayer object is destroyed, all of its inventory is destroyed too and will send parasite packets with fake informations (like item removal)
- [x] Prevent a player to modify another player's inventory

### CODE REVIEW

**Comment by tornac1234** [comment]
Cool thing to test is that RemotePlayerIdentifier is a MonoBehaviour so you can add anything inside to be related to the RemotePlayer instance. For example to test this PR you can add the following lines and use the debugger to verify if destroying the player will no longer send fake packets.
```c#
public void DestroyNow()
{
    Destroy(gameObject);
}
```

File: Nitrox.Assets.Subnautica/LanguageFiles/en.json [hunk 1]
```diff
@@ -64,5 +64,6 @@
     "Nitrox_WaitingUserInput": "Waiting for user input…",
     "Nitrox_WorldSettling": "Awaiting World Settling…",
     "Nitrox_ThankForPlaying": "Thanks for using Nitrox!",
-    "Nitrox_Failure": "An error occured"
+    "Nitrox_Failure": "An error occured",
+    "RemotePlayerObstacle": "Another player's either inside or too close to the deconstructable target."
```
Comment (Measurity) [inline Nitrox.Assets.Subnautica/LanguageFiles/en.json:68]: All other keys start with `Nitrox_`
```suggestion
    "Nitrox_RemotePlayerObstacle": "Another player's either inside or too close to the deconstructable target."
```

File: NitroxClient/GameLogic/PlayerLogic/RemotePlayerIdentifier.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+/// <summary>
+/// Attached to a RemotePlayer. Useful to determine that this script's GameObject is in the root of a RemotePlayer.
+/// </summary>
+public class RemotePlayerIdentifier : MonoBehaviour, IObstacle
+{
+    public RemotePlayer RemotePlayer;
+
+    public bool CanDeconstruct(out string reason)
+    {
+        reason = Language.main.Get("RemotePlayerObstacle");
```
Comment (Measurity) [inline NitroxClient/GameLogic/PlayerLogic/RemotePlayerIdentifier.cs:14]:
```suggestion
        reason = Language.main.Get("Nitrox_RemotePlayerObstacle");
```

File: NitroxClient/GameLogic/RemotePlayer.cs [hunk 3]
```diff
@@ -283,6 +287,45 @@ private void UpdateEquipmentVisibility()
...
+                // This layer lets us have a collider as a trigger without preventing its detection as an obstacle
+                Body.layer = LayerID.Useable;
+                Collider = Body.AddComponent<CapsuleCollider>();
+                
+                Collider.center = Vector3.zero;
+                Collider.radius = refCollider.radius;
+                Collider.direction = refCollider.direction;
+                Collider.contactOffset = refCollider.contactOffset;
+                Collider.isTrigger = true;
+            }
```
Comment (Jannify) [inline NitroxClient/GameLogic/RemotePlayer.cs:309]: We probably should log if it fails the if

File: NitroxClient/GameLogic/RemotePlayer.cs [hunk 4]
```diff
@@ -283,6 +287,45 @@ private void UpdateEquipmentVisibility()
...
+        }
+
+        /// <summary>
+        /// Change two parameters of the collider depending on the state of the player
+        /// </summary>
+        private void UpdateCollider(AnimChangeType type, AnimChangeState state)
+        {
+            if (type == AnimChangeType.UNDERWATER && state == AnimChangeState.ON)
```
Comment (Jannify) [inline NitroxClient/GameLogic/RemotePlayer.cs:317]: That way if the type is not UNDERWATER we always jump into the else, is this intended?

File: NitroxClient/GameLogic/PlayerLogic/RemotePlayerIdentifier.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
+﻿using UnityEngine;
+
+namespace NitroxClient.GameLogic.PlayerLogic;
+
+/// <summary>
+/// Attached to a RemotePlayer. Useful to determine that this script's GameObject is in the root of a RemotePlayer.
+/// </summary>
```
Comment (Measurity) [inline NitroxClient/GameLogic/PlayerLogic/RemotePlayerIdentifier.cs:7]: Might be good to explain what a root object is, for new devs especially.
```suggestion
/// <summary>
/// Attached to a RemotePlayer. Useful to determine that this script's GameObject is in the root of a RemotePlayer.
/// </summary>
/// <remarks>
/// The root of an object is defined as the top most GameObject as when an object hierarchy first was spawned in. Either from a prefab, or in Nitrox' case, a cloned root game object.
/// </remarks>
```

File: NitroxPatcher/Patches/Dynamic/Base_OnPreDestroy_Patch.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+public class Base_OnPreDestroy_Patch : NitroxPatch, IDynamicPatch
+{
+    private static readonly MethodInfo TARGET_METHOD = Reflect.Method((Base t) => t.OnPreDestroy());
+
+    public static void Prefix(Base __instance)
+    {
+        foreach (RemotePlayerIdentifier remotePlayerIdentifier in __instance.GetComponentsInChildren<RemotePlayerIdentifier>(true))
+        {
+            remotePlayerIdentifier.RemotePlayer.Detach();
+            SkyEnvironmentChanged.Broadcast(remotePlayerIdentifier.RemotePlayer.Body, (GameObject)null);
```
Comment (Jannify) [inline NitroxPatcher/Patches/Dynamic/Base_OnPreDestroy_Patch.cs:21]: Minor: What about moving this line into Detach()?

File: NitroxPatcher/Patches/Dynamic/Utils_GetEntityRoot_Patch.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+{
+    private static readonly MethodInfo TARGET_METHOD = Reflect.Method(() => UWE.Utils.GetEntityRoot(default));
+
+    public static bool Prefix(GameObject go, ref GameObject __result)
+    {
+        if (go.TryGetComponent(out RemotePlayerIdentifier remotePlayerIdentifier) || go.TryGetComponentInParent(out remotePlayerIdentifier))
+        {
+            __result = remotePlayerIdentifier.gameObject;
+        }
+        return __result == null;
```
Comment (Jannify) [inline NitroxPatcher/Patches/Dynamic/Utils_GetEntityRoot_Patch.cs:23]: Would be better to return true/false to not compare gameobjects directly to null.

**Review by Jannify — CHANGES_REQUESTED** [review]

File: NitroxClient/GameLogic/RemotePlayer.cs (hunk 4 above)
Comment (tornac1234) [reply NitroxClient/GameLogic/RemotePlayer.cs:317]: yes, I'll modify the comment above to explain that there are two collisions behaviours depending on whether or not the player is underwater

File: NitroxPatcher/Patches/Dynamic/Base_OnPreDestroy_Patch.cs (hunk 6 above)
Comment (tornac1234) [reply NitroxPatcher/Patches/Dynamic/Base_OnPreDestroy_Patch.cs:21]: In this case, I'll also move the same behaviour (but adapted to the furnished component) in Attach()

**Comment by dartasen** [comment]
Can probably fix #1841

**Comment by tornac1234** [comment]
> Can probably fix #1841

I could reproduce the duplicate items even alone in the server, so I'm pretty sure that this PR won't fix it

**Comment by tornac1234** [comment]
Tests proved it worked as intended

**Review by Jannify — APPROVED** [review]

**Review by dartasen — APPROVED** [review]

**Pull request merged**